spring/boot2-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java

```diff
@@ -116,144 +79,85 @@
+        return write(Flux.from(body));
     }
 
     @Override
-    public void beforeCommit(Supplier<? extends Mono<Void>> action) {
-        commitActions.add(action);
+    protected Mono<Void> writeAndFlushWithInternal(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+        // Prefetch 1 message because Armeria's HttpResponseSubscriber consumes messages one by one.
+        return write(Flux.from(body).concatMap(Flux::from, 1));
     }
 
     @Override
-    public boolean isCommitted() {
-        return state != State.NEW;
+    protected void applyStatusCode() {
+        final HttpStatus httpStatus = getStatusCode();
+        if (httpStatus != null) {
+            armeriaHeaders.status(httpStatus.value());
+        } else {
+            // If there is no status code specified, set 200 OK by default.
+            armeriaHeaders.status(com.linecorp.armeria.common.HttpStatus.OK);
+        }
     }
 
     @Override
-    public Mono<Void> writeWith(Publisher<? extends DataBuffer> body) {
-        return doCommit(() -> write(Flux.from(body)));
+    protected void applyHeaders() {
+        getHeaders().forEach((name, values) -> armeriaHeaders.add(HttpHeaderNames.of(name), values));
     }
 
     @Override
-    public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
-        // Prefetch 1 message because Armeria's HttpRequestSubscriber consumes messages one by one.
-        return write(Flux.from(body).concatMap(Flux::from, 1));
-    }
-
-    private Mono<Void> write(Flux<? extends DataBuffer> publisher) {
-        return Mono.deferContextual(contextView -> {
-            final HttpResponse response = HttpResponse.of(
-                    new HttpResponseProcessor(ctx.eventLoop(), buildResponseHeaders(),
-                                              publisher.map(factoryWrapper::toHttpData), contextView));
-            future.complete(response);
-            return Mono.fromFuture(response.whenComplete())
-                       .onErrorResume(cause -> cause instanceof CancelledSubscriptionException ||
-                                               cause instanceof AbortedStreamException,
-                                      cause -> Mono.empty());
-        });
-    }
-
-    private ResponseHeaders buildResponseHeaders() {
-        if (!armeriaHeaders.contains(HttpHeaderNames.STATUS)) {
-            // If there is no status code specified, set 200 OK by default.
-            armeriaHeaders.status(com.linecorp.armeria.common.HttpStatus.OK);
+    protected void applyCookies() {
+        final List<String> cookieValues =
+                getCookies().values().stream()
+                            .flatMap(Collection::stream)
+                            .map(ArmeriaServerHttpResponse::toSetCookie)
+                            .collect(toImmutableList());
+        if (!cookieValues.isEmpty()) {
+            armeriaHeaders.add(HttpHeaderNames.SET_COOKIE, cookieValues);
```

Comment:
nit: How about making `ArmeriaServerHttpResponse.toSetCookie()` return `Cookie` rather than `String` so that we use a shortcut method?
```java
armeriaHeaders.cookies(cookies);
```

Reply:
I knew that `ArmeriaServerHttpResponse.toSetCookie()` encodes a `Cookie` without the `strict` validation. 
Please ignore this review. :-)